Reduce the amount of repeating code in CT

gentoo-root opened the pull request:
Pure refactoring, no behavioral change.

gentoo-root committed:
bpf: Use ct_extract_ports4 for ct_lookup4

The switch in ct_lookup4 is almost identical to ct_extract_ports4. Reuse
the existing function with a slight modification.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Move ICMP_FRAG_NEEDED handling to a separate function

ICMP_FRAG_NEEDED handling in snat_v4_rev_nat is pretty self-contained,
and three indentation levels can easily be saved by moving this code
into a separate function with just slight modifications.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

qmonnet reviewed (approved):
Looks good, thanks!

Nit: In 2nd commit, some comment blocks could be re-wrapped now that they have less left-indent.

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1064,10 +1061,85 @@ snat_v4_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target, __s8 *
+static __always_inline __maybe_unused int
+snat_v4_rev_nat_handle_icmp_frag_needed(struct __ctx_buff *ctx, __u64 off)
+{
+	struct ipv4_ct_tuple tuple = {};
+	struct ipv4_nat_entry *state;
+	struct iphdr iphdr;
```
Just for context - when initially adding this code (a short while back), there was concern whether we should share the CT tuple (for processing the inner & outer packet), or have it cleanly separated. And how it would impact complexity.

gentoo-root replied:
Thanks, I wasn't aware of this context. What was the conclusion of that discussion? Shall I pass a pointer to tuple_buf to this function to reuse the one from the caller? Or there wasn't a clear conclusion?

julianwiedmann replied:
>What was the conclusion of that discussion?

We tried to share it as much as possible - but I don't think there was a final experiment on the latest code whether having it fully split up would trigger the verifier. So if we can do a clean split and still pass the verifier, I think that would be preferable.

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1064,10 +1061,85 @@ snat_v4_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target, __s8 *
+	if (IS_ERR(ret))
+		return ret;
+
+	/* Switch back to the outer header. */
+	tuple.nexthdr = IPPROTO_ICMP;
+
+	return snat_v4_rewrite_ingress(ctx, &tuple, state, off);
+}
```
I'm slightly concerned that this part (how we NAT the outer headers) will over time drift apart from what we do in the "main" path. How bad would it be to only handle the inner SNAT here, and return to the main function for handling the outer SNAT?

gentoo-root replied:
You mean just this function call `return snat_v4_rewrite_ingress(ctx, &tuple, state, off);`? I don't see what can drift here, rewriting the outer headers is already contained in a separate function, so that it can be reused from two different flows.

I can return 0 from here and `goto rewrite_ingress` in the caller on non-errors, but that's going to be uglier, and I don't see much point. Please tell me what you think.

julianwiedmann replied:
> You mean just this function call `return snat_v4_rewrite_ingress(ctx, &tuple, state, off);`? I don't see what can drift here, rewriting the outer headers is already contained in a separate function, so that it can be reused from two different flows.
> 

The bad scenario would be that at some point we only modify the main path, without realizing that there's a second user buried somewhere else.

> I can return 0 from here and `goto rewrite_ingress` in the caller on non-errors, but that's going to be uglier, and I don't see much point. Please tell me what you think.

Then let's not change it and keep your current approach :)

julianwiedmann reviewed (commented):
:100: on the `ct_extract_ports4()` patch!

I'm less convinced that the `ICMP_FRAG_NEEDED` patch is truly making it easier to understand things, but not feeling strongly about it either. But should we give the `ICMPV6_PKT_TOOBIG` part the same treatment then?

gentoo-root reviewed (commented):
> should we give the ICMPV6_PKT_TOOBIG part the same treatment then?

Good idea, I didn't notice IPv6 had a similar flow, because I mainly worked on IPv4 code for my bugfix. Let me do that too.

BTW, why don't we have ICMPV6_PKT_TOOBIG in snat_v6_nat, but only have it in revNAT?

gentoo-root committed:
bpf: Move ICMPV6_PKT_TOOBIG handling to a separate function

ICMPV6_PKT_TOOBIG handling in snat_v4_rev_nat is pretty self-contained,
and two indentation levels can easily be saved by moving this code into
a separate function with just slight modifications.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>
